File: internal/rss/rss.go
package rss

import (
	"encoding/xml"
	"os"
	"path/filepath"
	"time"

	"github.com/93percentai/sitemaker/internal/config"
	"github.com/93percentai/sitemaker/internal/content"
)

type rssRoot struct {
	XMLName xml.Name `xml:"rss"`
	Version string   `xml:"version,attr"`
	Channel channel  `xml:"channel"`
}

type channel struct {
	Title       string `xml:"title"`
	Link        string `xml:"link"`
	Description string `xml:"description"`
	Language    string `xml:"language"`
	LastBuild   string `xml:"lastBuildDate"`
	Items       []item `xml:"item"`
}

type item struct {
	Title       string   `xml:"title"`
	Link        string   `xml:"link"`
	Description string   `xml:"description"`
	PubDate     string   `xml:"pubDate"`
	GUID        string   `xml:"guid"`
	Categories  []string `xml:"category"`
}

func Generate(site *content.Site, cfg config.Config) error {
	feed := rssRoot{
		Version: "2.0",
		Channel: channel{
			Title:       cfg.Site.Title,
			Link:        cfg.Site.BaseURL,
			Description: cfg.Site.Description,
			Language:    cfg.Site.Language,
			LastBuild:   time.Now().Format(time.RFC1123Z),
		},
	}

	limit := 20
	if len(site.Posts) < limit {
		limit = len(site.Posts)
	}

	for _, post := range site.Posts[:limit] {
		fullURL := cfg.Site.BaseURL + post.URL
		feed.Channel.Items = append(feed.Channel.Items, item{
			Title:       post.Title,
			Link:        fullURL,
			Description: string(post.HTMLContent),
			PubDate:     post.Date.Format(time.RFC1123Z),
			GUID:        fullURL,
			Categories:  post.Tags,
		})
	}

	data, err := xml.MarshalIndent(feed, "", "  ")
	if err != nil {
		return err
	}

	output := []byte(xml.Header)
	output = append(output, data...)

	outDir := cfg.Build.OutputDir
	if err := os.MkdirAll(outDir, 0755); err != nil {
		return err
	}

	if err := os.WriteFile(filepath.Join(outDir, "rss.xml"), output, 0644); err != nil {
		return err
	}

	return os.WriteFile(filepath.Join(outDir, "feed.xml"), output, 0644)
}

File: src/rss.ts
import fs from "node:fs";
import path from "node:path";
import { Config } from "./config.js";
import { Site } from "./content.js";

export function generateRSS(site: Site, cfg: Config): void {
  const posts = site.posts.slice(0, 20);

  let xml = `<?xml version="1.0" encoding="UTF-8"?>
<rss version="2.0" xmlns:atom="http://www.w3.org/2005/Atom">
<channel>
  <title>${escXml(cfg.site.title)}</title>
  <link>${escXml(cfg.site.base_url)}</link>
  <description>${escXml(cfg.site.description)}</description>
  <language>${escXml(cfg.site.language)}</language>
  <lastBuildDate>${new Date().toUTCString()}</lastBuildDate>
  <atom:link href="${escXml(cfg.site.base_url)}rss.xml" rel="self" type="application/rss+xml"/>
`;

  for (const post of posts) {
    const fullURL = cfg.site.base_url.replace(/\/$/, "") + post.url;
    xml += `  <item>
    <title>${escXml(post.title)}</title>
    <link>${escXml(fullURL)}</link>
    <description><![CDATA[${post.htmlContent}]]></description>
    <pubDate>${post.date?.toUTCString() ?? ""}</pubDate>
    <guid>${escXml(fullURL)}</guid>
`;
    for (const tag of post.tags) {
      xml += `    <category>${escXml(tag)}</category>\n`;
    }
    xml += `  </item>\n`;
  }

  xml += `</channel>\n</rss>\n`;

  const outDir = cfg.build.output_dir;
  fs.mkdirSync(outDir, { recursive: true });
  fs.writeFileSync(path.join(outDir, "rss.xml"), xml);
  fs.writeFileSync(path.join(outDir, "feed.xml"), xml);
}

function escXml(s: string): string {
  return s
    .replace(/&/g, "&amp;")
    .replace(/</g, "&lt;")
    .replace(/>/g, "&gt;")
    .replace(/"/g, "&quot;");
}
